variables-vm.tf

```
@@ -147,6 +147,12 @@ variable "public_ip_enabled" {
   nullable    = false
 }
 
+variable "public_ip_id" {
+  description = "ID of the Public IP to attach to the Virtual Machine. if not provided, a new Public IP will be created."
+  type        = string
+  default     = null
+}
+
```

Use an object so that the `for_each` can be evaluated even if ID is dynamic. See what's done for `availability_set` variable for example.

```suggestion
variable "public_ip {
  description = "Public IP to attach to the Virtual Machine. if not provided, a new Public IP will be created."
  type = object({
    id = string
  })
  default     = null
}
```

---

done